keps/components/init-agent/0004_init-agent.md

```diff
@@ -0,0 +1,90 @@
+    1. Init-agent creates an ApiBinding mapping to the aforementioned APIExport and creates it in the newly created end-user workspace
+    1. Init-agent waits for all requested APIs to become ready
+    1. Init-agent creates any objects requested in the InitTarget and creates any missing namespaces
+1. Init-agent removes its initializer from the list
+1. (Assuming init-agents initializer was the last initializer): The end-user workspace gets marked as ready and is available to receive requests
+
+## CRDs
+
+In order for the init-agent to support different WorkspaceTypes, we need a mechanism to precisely tell it which APIs and Objects it should create upstream. We refer to this as the `InitTarget`:
+
+```yaml
+apiVersion: initialization.kcp.io/v1
+kind: InitTarget
+spec:
+  workspaceTypeRef:
+    name: myworkspacetype
+    path: root
+  sources:
+    - type: initTemplate
+      name: myTemplate
+      namespace: myNamespace
+```
+
+For the initial version of the agent, `initTemplate` is the only type of source for `.spec.sources`. We purposefully keep the API flexible in this regard, so future types could be added (e.g. templateWorkspace).
+For any objects mentioned in `.spec.sources`, the init-agent will automatically ensure that the corresponding APIs are available in the workspace via an APIExport+APIBinding. The init-agent will create any missing namespaces. The `InitTarget` is intended to be used in 1:n relationship with workspaces. As a result, the init-agent should log any issues while synching directly instead of pasting them as events in the InitTarget. This also helps to not overfill the object on large installations.
+
+In order to allow for full custom templating of all fields - including fields like `.metadata.name` - we need to have a wrapping Template resource. As a templating engine, Go Template in conjunctions with sprig is used:
+
+```yaml
+apiVersion: initialization.kcp.io/v1
+kind: InitTemplate
+metdata:
+  name: myTemplate
+spec:
+  # TODO (implementation detail): Do we want this to be free or structured text
+  text: |
+    apiVersion: v1
+    kind: Secret
+    metadata:
+      name: {{ .upstream.workspaceName }}-my-secret
+    data:
+      password: {{ sprig.randAlphaNum(10) }}
+```
+
+## Considered Downsides
+
+### API CRD maintenance
+
+By design, InitTemplates abstract away the need to have the API available on the cluster in order to make full templating possible and skip any OpenAPISchema validation. Specifically, InitTemplates can be applied successfully, even if no API matching the object described in `.spec.text` exists. Additionally any object created downstream may be invalid. However given our current requirement of full-text templating, we don't see an elegant way to work around this. What is possible theoretically is that a webhook would deny any InitTemplates whose inferred APIs are not yet available. This should be considered separately of the main controller of this agent.
```

Comment:
> Specifically, InitTemplates can be applied successfully

This means they can be created in the backing workspace, right? "Applying" sounds like "rendering into a new workspace", which should definitely not work if APIs are missing.